data-tx/src/main/java/io/micronaut/transaction/support/TransactionSynchronizationManager.java

```diff
@@ -474,93 +543,145 @@ public static void setActualTransactionActive(boolean active) {
+     * @param <T> The supplied type
+     * @return The suppler return value
+     * @author Denis Stepanov
+     * @since 3.4.0
+     */
+    @Internal
+    public static <T> T withState(@Nullable State state, Supplier<T> supplier) {
+        if (state == null) {
+            return supplier.get();
+        }
+        State previousState = getState();
+        try {
+            setState(state);
+            return supplier.get();
+        } finally {
+            setState(previousState);
+        }
+    }
+
+    /**
+     * Decorate the supplier with possible propagated state in thread-local.
+     *
+     * It's used to propagated TX state down to the async functions.
+     *
+     * @param supplier The supplier to be decorated
+     * @param <T> The supplied type
+     * @return The decorated supplier
+     * @author Denis Stepanov
+     * @since 3.4.0
+     */
+    @Internal
+    public static <T> Supplier<T> decorateToPropagateState(Supplier<T> supplier) {
+        State state = STATE.get();
+        if (state == null) {
+            return supplier;
+        }
+        return () -> withState(state, supplier);
+    }
+
     /**
      * The synchronization state.
      */
     @Internal
     public interface State {
     }
 
-    private static final class CopyState implements State {
-        private final Map<Object, Object> resources;
-        private final Set<TransactionSynchronization> synchronizations;
-        private final String currentTransactionName;
-        private final Boolean currentTransactionReadOnlyStatus;
-        private final TransactionDefinition.Isolation currentTransactionIsolationLevel;
-        private final Boolean actualTransactionActive;
-
-        private CopyState(Map<Object, Object> resources,
-                          Set<TransactionSynchronization> synchronizations,
-                          String currentTransactionName,
-                          Boolean currentTransactionReadOnlyStatus,
-                          TransactionDefinition.Isolation currentTransactionIsolationLevel,
-                          Boolean actualTransactionActive) {
-            this.resources = resources == null ? null : new HashMap<>(resources);
-            this.synchronizations = synchronizations == null ? null : new LinkedHashSet<>(synchronizations);
-            this.currentTransactionName = currentTransactionName;
-            this.currentTransactionReadOnlyStatus = currentTransactionReadOnlyStatus;
-            this.currentTransactionIsolationLevel = currentTransactionIsolationLevel;
-            this.actualTransactionActive = actualTransactionActive;
+    /**
+     * The copy-state of the thread-local values.
+     *
+     * @author Denis Stepanov
+     * @since 3.4.0
+     */
+    private static final class MutableState implements State {
+        private final Map<Object, Object> resources = new HashMap<>();
+        private final Map<Object, SynchronousTransactionState> states = new HashMap<>();
```

Comment:
Can we set sizes on these hash maps? There is unlikely to be many of these around at any one time